acvm-repo/acvm/tests/solver.rs

```diff
@@ -727,6 +726,200 @@ fn memory_operations() {
+    ];
+    let bn254_fr: Vec<u8> = vec![
+        1, 0, 0, 240, 147, 245, 225, 67, 145, 112, 185, 121, 72, 232, 51, 40, 93, 88, 129, 129,
+        182, 69, 80, 184, 41, 160, 49, 225, 114, 78, 100, 48,
+    ];
+    let secpk1_fr: Vec<u8> = vec![
+        0x41, 0x41, 0x36, 0xD0, 0x8C, 0x5E, 0xD2, 0xBF, 0x3B, 0xA0, 0x48, 0xAF, 0xE6, 0xDC, 0xAE,
+        0xBA, 0xFE, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xFF,
+    ];
+    let secpk1_fq: Vec<u8> = vec![
+        0x2F, 0xFC, 0xFF, 0xFF, 0xFE, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xFF, 0xFF,
+    ];
+    let secpr1_fq: Vec<u8> = vec![
+        0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0xFF, 0xFF,
+        0xFF, 0xFF,
+    ];
+    let secpr1_fr: Vec<u8> = vec![
+        81, 37, 99, 252, 194, 202, 185, 243, 132, 158, 23, 167, 173, 250, 230, 188, 255, 255, 255,
+        255, 255, 255, 255, 255, 0, 0, 0, 0, 255, 255, 255, 255,
+    ];
+
+    vec![bn254_fq, bn254_fr, secpk1_fr, secpk1_fq, secpr1_fq, secpr1_fr]
+}
+
+/// Whether to use a FunctionInput::constant or FunctionInput::witness:
+///
+/// (value, use_constant)
+type ConstantOrWitness = (FieldElement, bool);
+
+// For each ConstantOrWitness,
+// - If use_constant, then convert to a FunctionInput::constant
+// - Otherwise, convert to FunctionInput::witness
+//   + With the Witness index as (input_index + offset)
+//
+// Both use FieldElement::max_num_bits as the number of bits.
+fn constant_or_witness_to_function_inputs(
+    xs: Vec<ConstantOrWitness>,
+    offset: usize,
+) -> Vec<FunctionInput<FieldElement>> {
+    xs.into_iter()
+        .enumerate()
+        .map(|(i, (x, use_constant))| {
+            if use_constant {
+                FunctionInput::constant(x, FieldElement::max_num_bits())
+            } else {
+                FunctionInput::witness(Witness((i + offset) as u32), FieldElement::max_num_bits())
+            }
+        })
+        .collect()
+}
+
+// Convert ConstantOrWitness's back to FieldElement's by dropping the bool's
+fn drop_use_constant(input: &[ConstantOrWitness]) -> Vec<FieldElement> {
+    input.iter().map(|x| x.0).collect()
+}
+
+prop_compose! {
+    fn bigint_with_modulus()(modulus in select(allowed_bigint_moduli()))
+        (input in proptest::collection::vec(any::<(u8, bool)>(), modulus.len()), modulus in Just(modulus))
+        -> (Vec<ConstantOrWitness>, Vec<u8>) {
+        let input = input.into_iter().zip(modulus.iter()).map(|((x, use_constant), modulus_byte)| {
+            (FieldElement::from(x.clamp(0, *modulus_byte) as u128), use_constant)
+        }).collect();
+        (input, modulus)
+    }
+}
+
+prop_compose! {
+    fn bigint_pair_with_modulus()(input_modulus in bigint_with_modulus())
```

Comment:
Unfortunately not: `proptest!` supports `(x, y) in gen()`, but `prop_compose!` doesn't

Reply:
Can we split up the tuple in the function code then?

Reply:
Yes, good point: I started using `.0` `.1` because they were required for the "code as parameters" macro stuff, but can switch back to normal tuples in the body 👍